### DISCUSSION TITLE
Adding Secondary DNS or MQTT Host Fallback

### DISCUSSION THREAD

**Discussion by Gosokan** [arendst/Tasmota#11921] [Ideas]
It would be nice if it was possible to add secondary dns or mqtt host fallback.
So when the main dns server experiences downtime, the tasmota device can connect to secondary mqtt broker.

**Comment by barbudor** [comment]
Hi @Gosokan 
I'm not 100% sure but when using DHCP, it should work to receive 2 DNS servers from the DHCP server. Have you tried it ?
When using static IP, it seems that the Arduino core could support 2 DNS servers but some work would be required on Tasmota with -from my point of view- little advantage.

But as a more general point, what would be the real use-case of having multiple DNS server at the ESP level ?
You should be able to have your router (or your home server) doing the balancing automatically.


With regards to MQTT, I don't see the need for supporting multiple MQTT servers as this could lead to strange situations with half of the devices connected to one server and the other half connected to the other server.
If you need a resilient MQTT infrastructure, work on making your infrastructure resilient :
- 1st, insure that your MQTT server cannot be down (UPS, backups, no SD card, ...)
- 2nd if you need more, use 2 MQTT servers with a virtual IP so you are sur only a single 1 is active for all clients at a time

**Reply by Gosokan** [reply]
The dhcp server on the mikrotik router has been set to use 2 dns servers:
![image](https://user-images.githubusercontent.com/13587422/116359505-1129a700-a829-11eb-896b-dffae7c16a4d.png)

But in Tasmota information page, only primary dns are detected. 
![image](https://user-images.githubusercontent.com/13587422/116359640-3b7b6480-a829-11eb-8630-9d579926115f.png)

The solution I currently use is to set the hostname for both mqtt brokers with the same name, but with different IP addresses on each dns server.

There is little chance of power interruptions. Because I am using a large battery backup (12V 45Ah) which can last more than 1 day for running Raspbery Pi 4 , mikrotik, and access point without mains electricity. But 

But this raspberry pi definitely needs to restart for software updates and security fixes, or other problems. So if Tasmota can use more than 1 dns server, then I don't need to worry anymore about the inaccessible mqtt broker.

**Reply by sfromis** [reply]
Looks like you want automatic failover to another MQTT broker, but due to retained messages, simple DNS based failover will not work properly for MQTT; the other broker would lack current state.

**Reply by barbudor** [reply]
I know that Tasmota display only 1 DNS server.
But Arduino core code still seems to be able to receive 2 from DHCP.
That should be easily testable as you have control over 2 own DNS server.

**Reply by rapi3** [reply]
It wold be nice to have a setting for second MQTT server.
I am using 2 MQTT servers, one as a bridge so my topics are synchronized between MQTT servers but in case tasmota device lose connection to MQTT server it is a problem.

**Comment by joba-1** [comment]
If tasmota devices lose network connectivity, a second MQTT broker wont help, so I'm ignoring that case and what you want is the MQTT service to be highly available. 
Syncing retained topics between MQTT brokers is the first obstacle. It can be solved, but for most usecases tasmota wont care about retained topics anyways so you could ignore that. Most of the time any freshly started MQTT broker would do.
Second problem is the single IP tasmota (and other MQTT clients) use to connect to the service. It could be solved by tasmota (and all other clients, including your home automation software) using a second IP to connect to, but it can be solved centrally in one place by adding a second IP to your primary MQTT server. If the primary server goes down you can add this dead second IP to your other MQTT server and clients will (re)connect there. For planned maintenance or after noticing the failure you can do that IP move manually. But there are also solutions that do this move automatically (heartbeat, pacemaker, corosync, ...).

**Reply by sfromis** [reply]
While the operation of Tasmota would indeed normally not really be affected by the error of missing retained messages, operation of backend platforms like Home Assistant could be severely hit by the lack of config data and LWT.
